pkg/ipalloc/ipalloc.go

```diff
@@ -0,0 +1,617 @@
+
+	start := big.NewInt(0)
+	start.SetBytes(a.start.AsSlice())
+
+	// The range .10 to .20 has 11 IPs, so calc is #available = (end-start+1) - #allocated
+	available = available.Sub(available, start)
+	available = available.Add(available, big.NewInt(1))
+	available = available.Sub(available, big.NewInt(int64(allocated)))
+
+	return allocated, available
+}
+
+// Range returns the start and stop IP of the allocator.
+func (a *HashAllocator[T]) Range() (from, to netip.Addr) {
+	return a.start, a.stop
+}
+
+// scheduleSort schedules a sort to happen with one second delay, if a sort is already scheduled we do nothing.
+// This coalesces what otherwise would be multiple sort operations if lots of allocations happen in a short
+// time. Sorting of available blocks improves memory usage in the long term but is not critical to the
+// operation of the allocator, and thus its alright to delay the sort.
+func (a *HashAllocator[T]) scheduleSort() {
+	if a.timer != nil {
+		return
+	}
+
+	a.timer = time.NewTimer(time.Second)
+	go func() {
+		<-a.timer.C
+		a.mu.Lock()
+		defer a.mu.Unlock()
+
+		sort.Sort(a.availableBlocks)
+		a.timer = nil
+	}()
+}
+
+// A slice of blocks, sorted by size from smallest to largest. But every element is also part of
+// a linked list ordered according to adjacency in the IP space.
+// When allocating, we always want to allocate from the smallest range so we reduce the amount
+// of blocks to keep track of. And the links are used for merging blocks.
+type availableBlockList []*linkedBlock
+
+func (abl availableBlockList) String() string {
+	var lines []string
+	lines = append(lines, "linked-block-list[")
+	for _, block := range abl {
+		lines = append(lines, block.String())
+	}
+	lines = append(lines, "]")
+	return strings.Join(lines, "\n")
+}
+
+// take removes `ip` from the list of available IPs by shrinking one of the blocks or by splitting a block
+// in two if the `ip` appears in the middle of a range.
+func (abl *availableBlockList) take(ip netip.Addr) error {
+	var (
+		availableBlock *linkedBlock
+		index          int
+	)
+
+	// NOTE this search can be improved by traversing the linked list O(n) search
+	for i, block := range *abl {
+		if block.within(ip) {
+			availableBlock = block
+			index = i
+			break
+		}
+	}
+	if availableBlock == nil {
+		return ErrInUse
+	}
+
+	// If its the first IP, we can shrink the block
+	if availableBlock.from == ip {
+		// If this is the last IP in the block, remove the block
+		if availableBlock.from == availableBlock.to {
+			abl.delete(index)
+			return nil
+		}
+
+		availableBlock.from = availableBlock.from.Next()
+		availableBlock.calcSize()
+		return nil
+	}
+
+	if availableBlock.to == ip {
+		availableBlock.to = availableBlock.to.Prev()
+		availableBlock.calcSize()
+		return nil
+	}
+
+	// At this point, it must be an IP in the middle of the block, so we need to split it
+
+	// Remove the current block from the list
+	*abl = slices.Delete(*abl, index, index+1)
```

Comment:
Yea, it could be. Though the list only contains pointers, so its just a copy of 8*#blockRightOfIndex amount of bytes, you would have to have things fragmented very heavily. But if we go for a heap to sort blocks this operation should become cheaper only a pointer update.